### PR TITLE
Support for bash script

### PR DESCRIPTION
**Pull request by mvsusp**
# Support for bash script

## before changes

```python
estimator = MXNET(entrypoint='train.py')
```
Script mode will:

1. Add a setup.py to the script if it doesn't exist (https://github.com/aws/sagemaker-containers/blob/d20115a8a2c7024b496a94700237c641b0bf453b/src/sagemaker_containers/_modules.py#L54)  
2. Pip installs the customer script as a python package, using the requirements.txt file if provided (https://github.com/aws/sagemaker-containers/blob/d20115a8a2c7024b496a94700237c641b0bf453b/src/sagemaker_containers/_modules.py#L116)
3. Write the Training Environment as environment variables (https://github.com/aws/sagemaker-containers/blob/master/src/sagemaker_containers/_modules.py#L322): the list of env vars is here (https://github.com/aws/sagemaker-containers#list-of-provided-environment-variables-by-sagemaker-containers)
4. Runs the python *module* as follow:  cmd = [python_executable(), '-m', module_name] + args  where *args* are customer's hyperparameters

### If the customer provides python script without setup.py:

```python
estimator = MXNET(entrypoint='train.py', source_dir='src')
```

Script mode will:

1. Copy src directory to /opt/ml/code
2. Write the Training Environment as environment variables (https://github.com/aws/sagemaker-containers/blob/master/src/sagemaker_containers/_modules.py#L322): the list of env vars is here (https://github.com/aws/sagemaker-containers#list-of-provided-environment-variables-by-sagemaker-containers)
3. add /opt/ml/code to PYTHONPATH
4. os.chdir('/opt/ml/code')
5. Runs the python *script* as follow:  cmd = [python_executable(), script_name] + args  where *args* are customers hyperparameters

### If customer provides a non python entrypoint

```python
estimator = MXNET(entrypoint='train.sh', source_dir='src')
```
or

```python
estimator = MXNET(entrypoint='train', source_dir='src')
```

Script mode will:

1. Copy src directory to /opt/ml/code
2. Write the Training Environment as environment variables (https://github.com/aws/sagemaker-containers/blob/master/src/sagemaker_containers/_modules.py#L322): the list of env vars is here (https://github.com/aws/sagemaker-containers#list-of-provided-environment-variables-by-sagemaker-containers)
3. add /opt/ml/code to PYTHONPATH
4. os.chdir('/opt/ml/code')
5. Runs the *script* as follow:  cmd = [script_name] + args  where *args* are customers hyperparameters

### If customer provides python script with setup.py

Script mode will keep the current functionality

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker_containers/_modules.py [hunk 1]
```diff
@@ -133,64 +108,52 @@ def exists(name):  # type: (str) -> bool
...
-    # type: (str, str, bool) -> module
-    """Download, prepare and install a compressed tar file from S3 or local directory as a module.
+def download_and_install(uri, name, path):  # type: (str, str, str) -> None
+    """Download, prepare and install a compressed tar file from S3 or local directory as an entry point.
 
-    SageMaker Python SDK saves the user provided scripts as compressed tar files in S3
+    SageMaker Python SDK saves the user provided entry points as compressed tar files in S3
     https://github.com/aws/sagemaker-python-sdk.
 
     This function downloads this compressed file, if provided, and transforms it as a module, and installs it.
```
Comment (icywang86rui) [inline src/sagemaker_containers/_modules.py]: This is not always true anymore.

File: src/sagemaker_containers/_modules.py [hunk 2]
```diff
@@ -223,23 +186,34 @@ def run(module_name, args=None, env_vars=None, wait=True):  # type: (str, list,
...
-    cmd = [python_executable(), '-m', module_name] + args
+    # TODO (mvsusp): module_name will be deprecated to entry_point_name in a follow up pr that will refactor _modules
+    entry_point_type = _entry_point_type(_env.code_dir, module_name)
+
+    if entry_point_type == _PYTHON_PACKAGE:
+        cmd = [python_executable(), '-m', module_name.replace('.py', '')] + args
+    elif entry_point_type == _PYTHON_PROGRAM:
+        cmd = [python_executable(), module_name] + args
+    else:
+        cmd = ['/bin/sh', '-c', 'chmod +x %s && ./%s %s' % (module_name, module_name, ' '.join(args))]
```
Comment (icywang86rui) [inline src/sagemaker_containers/_modules.py]: Is the chmod here necessary? Didn't we already handled this in prepare?

File: src/sagemaker_containers/_modules.py [hunk 3]
```diff
@@ -281,26 +258,27 @@ def import_module(uri, name=DEFAULT_MODULE_NAME, cache=True):  # type: (str, str
         six.reraise(_errors.ImportModuleError, _errors.ImportModuleError(e), sys.exc_info()[2])
 
 
-def run_module(uri, args, env_vars=None, name=DEFAULT_MODULE_NAME, cache=True, wait=True):
+def run_module(uri, args, env_vars=None, name=DEFAULT_MODULE_NAME, cache=None, wait=True):
```
Comment (icywang86rui) [inline src/sagemaker_containers/_modules.py]: Does this change break some of our framework containers?

File: src/sagemaker_containers/_modules.py [hunk 4]
```diff
@@ -317,6 +295,38 @@ def write_env_vars(env_vars=None):  # type: (dict) -> None
 
     """
     env_vars = env_vars or {}
+    env_vars['PYTHONPATH'] = ':'.join(sys.path)
 
     for name, value in env_vars.items():
         os.environ[name] = value
+
+
+_PYTHON_PACKAGE = 'PYTHON_PACKAGE'
```
Comment (icywang86rui) [inline src/sagemaker_containers/_modules.py]: Minor:
Make a enum for these values

File: src/sagemaker_containers/_trainer.py [hunk 5]
```diff
@@ -42,19 +42,23 @@ def train():
         # if the framework module is not defined.
         env = sagemaker_containers.training_env()
 
-        framework_name, entry_point_name = env.framework_module.split(':')
+        if env.framework_module:
+            framework_name, entry_point_name = env.framework_module.split(':')
```
Comment (icywang86rui) [inline src/sagemaker_containers/_trainer.py]: I would refactor this part under the 'if' into a function. YMMV

**Comment by icywang86rui** [comment]
Looks good to me. Do we want to add Leo or Nadia too the pr since they know more about this use case?

File: test/functional/test_download_and_import.py [hunk 6]
```diff
@@ -42,27 +42,20 @@ def erase_user_module():
...
-
-    assert module.validate()
-
-
-def test_import_module_with_local_script(user_module_name, tmpdir):
+def test_import_module_with_local_uri(user_module_name, tmpdir):
     tmp_code_dir = str(tmpdir)
 
-    test.UserModule(USER_SCRIPT).create_tmp_dir_with_files(tmp_code_dir)
+    test.UserModule(USER_SCRIPT).add_file(SETUP).create_tmp_dir_with_files(tmp_code_dir)
```
Comment (nadiaya) [inline test/functional/test_download_and_import.py]: Are customers now expected to provide setup file?

File: test/unit/test_mapping.py [hunk 7]
```diff
@@ -204,7 +204,7 @@ def test_env_vars_round_trip():
     assert env_vars['SM_HPS'] == '{"batch_size":64,"epochs":10,"loss":"SGD","precision":5.434322}'
     assert env_vars['SM_HP_PRECISION'] == '5.434322'
     assert env_vars['SM_RESOURCE_CONFIG'] == '{"current_host":"algo-1","hosts":["algo-1","algo-2","algo-3"]}'
-    assert env_vars['SM_MODULE_NAME'] == 'user_script'
+    assert env_vars['SM_USER_PROGRAM'] == 'user_script.py'
```
Comment (nadiaya) [inline test/unit/test_mapping.py]: isn't it a breaking change?
what if someone expects the old environment variable?

File: test/unit/test_trainer.py [hunk 8]
```diff
@@ -49,11 +53,23 @@ def success():
     _exit.assert_called_with(_trainer.SUCCESS_CODE)
 
 
+@patch('sagemaker_containers._modules.run_module')
+@patch('sagemaker_containers.training_env', new_callable=SriptTrainingEnv)
```
Comment (nadiaya) [inline test/unit/test_trainer.py]: ScriptTrainingEnv

**Pull request closed without merging**